### PR TITLE
Test Impact Analysis Divide by zero fix

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
Fixed potential divide by zero in the Test Impact Analysis Framework when outputting which test were selected.

This was causing the Test Impact Analysis Framework run to crash with return code 0xc0000094 which is the return code for integer divide by zero on Windows

The error can be seen on Jenkins at https://jenkins.build.o3de.org/blue/organizations/jenkins/O3DE/detail/development/4331/pipeline/819

```
[2023-05-16T21:48:21.363Z] BuildTargetDependencyGraph: Trace::Warning

[2023-05-16T21:48:33.592Z]  D:\workspace\o3de\Code\Tools\TestImpactFramework\Runtime\Common\Code\Include\Static\BuildTarget/Common/TestImpactBuildGraph.h(156): 'auto __cdecl TestImpact::BuildGraph<class TestImpact::NativeProductionTarget,class TestImpact::NativeTestTarget>::{ctor}::<lambda_2>::operator ()(const class AZStd::vector<class AZStd::basic_string<char,struct AZStd::char_traits<char>,class AZStd::allocator>,class AZStd::allocator> &,class AZStd::unordered_set<struct TestImpact::BuildGraphVertex<class TestImpact::NativeProductionTarget,class TestImpact::NativeTestTarget> const *,struct AZStd::hash<struct TestImpact::BuildGraphVertex<class TestImpact::NativeProductionTarget,class TestImpact::NativeTestTarget> const *>,struct AZStd::equal_to<struct TestImpact::B[2023-05-16 21:48:31,702][TIAF][ERROR] The test impact analysis runtime returned with error: '3221225620'.
```

Added detection to the Test Impact console utilities to only output ANSI color escapes if the terminal supports it.

### CODE REVIEW

**Review by jonawals — APPROVED** [review]
Thanks for doing this change, @lumberyard-employee-dm, great work!

File: Code/Framework/AzCore/AzCore/IO/AnsiTerminalUtils.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+namespace AZ::IO::Posix
+{
+    //! Turns of Virtual Terminal Processing on Windows
```
Comment (mbalfour-amzn) [inline Code/Framework/AzCore/AzCore/IO/AnsiTerminalUtils.h:13]:
```suggestion
    //! Turns off Virtual Terminal Processing on Windows
```

File: Code/Framework/AzCore/AzCore/IO/AnsiTerminalUtils.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    //! This allows ANSI escapes to be supported
+    //! https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences
+    void EnableVirtualTerminalProcessing(int fileDescriptor);
+    //! return true the terminal supports color ANSI escape codes
```
Comment (mbalfour-amzn) [inline Code/Framework/AzCore/AzCore/IO/AnsiTerminalUtils.h:17]:
```suggestion
    //! return true if the terminal supports color ANSI escape codes
```

File: Code/Framework/AzCore/Platform/Common/WinAPI/AzCore/IO/AnsiTerminalUtils_WinAPI.cpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        GetConsoleMode(fileHandle, &currentMode);
+        SetConsoleMode(fileHandle, currentMode | ENABLE_VIRTUAL_TERMINAL_PROCESSING);
+#endif
+    }
+
+    bool TerminalSupportsColor()
+    {
+        return true;
+    }
+    // return true if the file descriptor is refering to a terminal
```
Comment (mbalfour-amzn) [inline Code/Framework/AzCore/Platform/Common/WinAPI/AzCore/IO/AnsiTerminalUtils_WinAPI.cpp:28]:
```suggestion
    // return true if the file descriptor is referring to a terminal
```

File: Code/Tools/TestImpactFramework/Frontend/Console/Common/Code/Source/TestImpactTestSequenceNotificationHandler.cpp [hunk 4]
```diff
@@ -32,11 +35,18 @@ namespace TestImpact
...
+                    const float saving = (1.0f - float(numSelectedTests) / totalTests) * 100.0f;
 
-                std::cout << numSelectedTests << " tests selected, " << numDiscardedTests << " tests discarded (" << saving << "% test saving)\n";
-                std::cout << "Of which " << numExcludedTests << " tests have been excluded and " << numDraftedTests << " tests have been drafted.\n";
+                    std::cout << numSelectedTests << " tests selected, " << numDiscardedTests << " tests discarded (" << saving << "% test saving)\n";
+                    std::cout << "Of which " << numExcludedTests << " tests have been excluded and " << numDraftedTests << " tests have been drafted.\n";
+                }
+                else
+                {
+                    std::cout << "There are 0 total test\n";
```
Comment (mbalfour-amzn) [inline Code/Tools/TestImpactFramework/Frontend/Console/Common/Code/Source/TestImpactTestSequenceNotificationHandler.cpp:48]:
```suggestion
                    std::cout << "There are 0 total tests\n";
```

**Review by mbalfour-amzn — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/IO/AnsiTerminalUtils.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    bool TerminalSupportsColor();
+    //! return if the open file descriptor is a terminal
```
Comment (spham-amzn) [inline Code/Framework/AzCore/AzCore/IO/AnsiTerminalUtils.h:19]:
```suggestion
    //! returns true if the open file descriptor is a terminal
```

File: Code/Framework/AzCore/AzCore/IO/AnsiTerminalUtils.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    bool IsAtty(int fileDescriptor);
+    //! return true if the file descriptor represents a terminal
+    //! and it supports ANSI escapes
+    bool SupportsAnsiEscapes(int fileDescriptor);
+}
```
Comment (spham-amzn) [inline Code/Framework/AzCore/AzCore/IO/AnsiTerminalUtils.h:24]:
```suggestion
    //! returns true if the file descriptor supports ansi escape characters
}
```

File: Code/Framework/AzCore/Platform/Common/UnixLike/AzCore/IO/AnsiTerminalUtils_UnixLike.cpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+                || termView == "tmux-256color"
+                || termView == "rxvt-unicode"
+                || termView == "rxvt-unicode-256color"
+                || termView == "linux";
+        }
+
+        return false;
+    }
+    // return true if the file descriptor is refering to a terminal
+    bool IsAtty(int fileDescriptor)
```
[the remark this answers is not in the document]
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/Platform/Common/UnixLike/AzCore/IO/AnsiTerminalUtils_UnixLike.cpp:40]: According to the O3DE coding standards, abbreviations are to be [lowercased](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#abbreviations).

Now for a correctness standpoint, I haven't found any rationale on why the method is called [isatty](https://pubs.opengroup.org/onlinepubs/009695299/functions/isatty.html) and not `istty`

**Review by spham-amzn — APPROVED** [review]

File: Code/Framework/AzCore/Platform/Common/UnixLike/AzCore/IO/AnsiTerminalUtils_UnixLike.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mbalfour-amzn) [reply Code/Framework/AzCore/Platform/Common/UnixLike/AzCore/IO/AnsiTerminalUtils_UnixLike.cpp:40]: The abbreviation though is "TTY", not "ATTY", so under the [function naming Camel-Case rules](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#function-names) each word would be capitalized - "Is A Tty". Kind of like "IsAScubaTank".

[the remark this answers is not in the document]
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/Platform/Common/UnixLike/AzCore/IO/AnsiTerminalUtils_UnixLike.cpp]: You win @mbalfour-amzn, it is has been renamed to "IsATty", though it does look awkward.

[the remark this answers is not in the document]
Comment (mbalfour-amzn) [reply Code/Framework/AzCore/Platform/Common/UnixLike/AzCore/IO/AnsiTerminalUtils_UnixLike.cpp]: lol doesn't _feel_ like a win. :( As an alternative, you could name it "IsTty" and drop the "a"...

**Pull request merged**